opentelemetry-sdk/src/metrics/internal/histogram.rs

```diff
@@ -37,6 +36,14 @@ where
             buckets: Mutex::new(Buckets::<T>::new(*count)),
         }
     }
+
+    fn clone_and_reset(&self, count: &usize) -> Self {
+        let mut current = self.buckets.lock().unwrap_or_else(|err| err.into_inner());
+        let cloned = replace(current.deref_mut(), Buckets::new(*count));
```

Comment:
This would lead to unnecessary allocation. The replacement with `Buckets::new` is only needed for the no attribute case as there could be newer updates trying to update the same tracker. For trackers with attributes, we could simply `mem::take` the buckets instead of creating a new set of buckets for each tracker.

Reply:
Yes, you're absolutely right we should be able to simply 'take' it from trackers with attributes, but at the moment we cant (or rather it would be super inefficient), because hashmap can contain several instances of same tracker (one with sorted attrib set, an N with not unsorted).
I actually plan to create revision for that as well (was afraid to do this as part as this revision, to reduce scope):)
Idea is to split this hashmap into two: one with sorted attributes  and another all attributes for faster lookup. This change would greatly improve collection performance, and by accident would solve #2093 as well:)

In summary, this is temporary minor inefficiency that will be solved:)

Reply:
> but at the moment we cant (or rather it would be super inefficient), because hashmap can contain several instances of same tracker

I didn't get that. Fixing this extra allocation here does not depend on the `Hashmap` containing two sets of the same attributes. There is a relatively simple fix to this:

1. Add a new method to the `Aggregator` trait and update `collect_and_reset` to call this new method for trackers with attributes

```rust
pub(crate) trait Aggregator {
    ...

    /// Return the current value
    fn get_value(&self, init: &Self::InitConfig) -> Self;
}

pub(crate) fn collect_and_reset<Res, MapFn>(&self, dest: &mut Vec<Res>, mut map_fn: MapFn)
where
    MapFn: FnMut(Vec<KeyValue>, A) -> Res,
{
    prepare_data(dest, self.count.load(Ordering::SeqCst));

    ...

    let mut seen = HashSet::new();
    for (attrs, tracker) in trackers.drain() {
        if seen.insert(Arc::as_ptr(&tracker)) {
            dest.push(map_fn(attrs, tracker.get_value(&self.config)));
        }
    }
}
```

2. Implement the new method in the `HistogramTracker` by not calling `Buckets::new`. Notice that we call `mem::take` here and not `mem::replace`.

```rust
fn get_value(&self, _init: &Self::InitConfig) -> Self
{
    let mut current = self.buckets.lock().unwrap_or_else(|err| err.into_inner());
    Self {
        buckets: Mutex::new(take(&mut current)),
    }
}
```

This should work.

Reply:
Agreed to come back to this in next PR.